### ISSUE TITLE
`odo logs` too slow for non-local clusters

### ISSUE THREAD

**Issue by dharmit** [redhat-developer/odo#5872]
## What versions of software are you using?

**Operating System:** all

**Output of `odo version`:** `main`

## How did you run odo exactly?
`odo logs` against a cluster that's not minikube/CRC. For example, one spun up using ClusterBot

## Actual behavior
Takes many seconds to show the logs

## Expected behavior
Take lesser time

## Any logs, error output, etc.?
Originally discussed at https://github.com/redhat-developer/odo/pull/5846#issuecomment-1164423486.

/kind bug
/area log

**Comment by rm3l** [comment]
> ## Expected behavior
> Take lesser time
>
To ensure we won't have regressions in the future, I think it could be great to include some kind of performance testing for this, with an acceptable threshold for what we mean by "lesser time". Relates to #5830, but for `odo logs`.

**Comment by dharmit** [comment]
Self-note:
* Look at client-go caching mentioned by Tomas
* By threshold, Armel referred to how long odo should wait till `odo logs` exits if it doesn't get a response it's looking for.

**Comment by dharmit** [comment]
Tasks pending on this issue:
* Address `odo logs --follow` issue explained in https://github.com/redhat-developer/odo/pull/5942#issuecomment-1187685258.
* In a further PR, modify the code to first fetch the resources most likely to have Pods of interest for odo. This would include Deployments, ReplicaSets and Pods itself. Check other resources after that.
* Use channels instead of sharing a slice+mutex in the go routines, as is the case here: https://github.com/redhat-developer/odo/blob/e7588e33ef412177cf050efe68fc2af0e16ac79b/pkg/kclient/all.go#L29-L57
* Anything else that we might be able to think of while working on the above.

We have discussed adding tests that would help us track the performance of `odo logs` so that we can know if it slows down in the future. This is to avoid regression, or at least finding it ourselves before it becomes a problem for the users. I prefer it be done in a separate issue than this.
